orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -963,229 +1131,461 @@ common::Status SplitGraph(Graph& graph,
-            }
-            return Status::OK();
-          });
-
-      if (upstream_nodes_output_index < 0) {
-        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Node with name: ", producer_node->Name(),
-                               " doesn't have an output node_arg with name ", id.node_arg_name);
-      }
-
-      size_t idx = static_cast<size_t>(upstream_nodes_output_index);
+    ORT_RETURN_IF_ERROR(
+      AddMetaTensors(current_stage, next_stage, graph, new_input_names,
+                     new_output_names, send_input_args, send_output_args,
+                     recv_input_args, recv_output_args, stage_to_rank));
 
-      // original node_arg pointer from the origin graph. This serves as the key in the
-      // updated_node_arg map and any reference for original node_arg name
-      auto* original_node_arg = producer_node->MutableOutputDefs()[idx];
-
-      // updated node_arg pointer from previous partition. This is the new arg_node the
-      // current inserted send node will take as input node_arg.
-      auto updated_node_arg = producer_node->MutableOutputDefs()[idx];
-      auto exiting_updated_node_arg = updated_node_args.find(original_node_arg);
-      if (exiting_updated_node_arg != updated_node_args.end()) {
-        updated_node_arg = exiting_updated_node_arg->second;
-      }
-      assert(updated_node_arg);
+    // Get all the node_args that need to be sent to the next stage.
+    auto& tensors_sent_in_forward = forward_messages.at(current_stage);
+    // TODO(jufranc): consider tensors sent by backward ops.
+    // auto& tensors_sent_in_backward = backward_messages.at(current_stage);
+    // auto& tensors_sent = current_stage + 1 == next_stage ? tensors_sent_in_forward : tensors_sent_in_backward;
 
-      send_input_args.push_back(updated_node_arg);
+    // Take care of tensors that need to be sent from one device to the other.
+    for (const NodeArg* arg : tensors_sent_in_forward) {
+      send_input_args.push_back(const_cast<NodeArg *>(arg));
 
-      auto dtype = original_node_arg->TypeAsProto()->tensor_type().elem_type();
+      // The tensor replica has been created in advance. We query it now
+      // because it will be one of the outputs of the receive node in this
+      // stage. We also need to add it to the graph.
+      NodeArg* new_receive_output = tensor_replicas.at(arg).at(next_stage);
+      recv_output_args.push_back(new_receive_output);
 
+      auto dtype = arg->TypeAsProto()->tensor_type().elem_type();
       element_types.add_ints(static_cast<int64_t>(dtype));
+    }
 
-      auto& new_receive_output = CreateNodeArg(graph, *updated_node_arg);
-      const auto old_shape = *(updated_node_arg->Shape());
-      new_receive_output.SetShape(old_shape);
-      recv_output_args.push_back(&new_receive_output);
+    // Take care of tensors that need to be forwarded.
+    for (auto& forwarding_entry : forwarded_tensors) {
+      const NodeArg* tensor = forwarding_entry.first;
```

Comment:
This function has many long for-loop's. Please consider to wrap them as functions.

Reply:
yes, it does. Will do that. Thanks